### PR TITLE
Value type surrogates + using regular constructor for records

### PR DESCRIPTION
**Pull request by mishun**
Changes include:

- Surrogates are emitted for value type records if necessary (i.e. if fields of type from {string, Array<>, Map<, >, Set<>, List<>} are present; previously such records caused runtime error)
- Complete factory method is emitted instead of separate setters in all other cases where factory method is needed.
- Factory method for records now prefers using standard record constructor over Runtime.Serialization.FormatterServices.GetUninitializedObject.

### CODE REVIEW

File: test/ProtoBuf.FSharp.Unit/TestRecordRoundtrip.fs
Comment (mvkara) [inline test/ProtoBuf.FSharp.Unit/TestRecordRoundtrip.fs]: Remove comment. You've worked around it by using a reference type surrogate.

File: test/ProtoBuf.FSharp.Unit/TestRecordRoundtrip.fs
Comment (mvkara) [inline test/ProtoBuf.FSharp.Unit/TestRecordRoundtrip.fs]: Remove comment.

File: src/ProtoBuf.FSharp/ZeroValues.fs
Comment (mvkara) [inline src/ProtoBuf.FSharp/ZeroValues.fs:17]: Minor point: Prefer one line breaks. I've created another branch (https://github.com/mvkara/fsharp-hashcollections/tree/develop) for you to merge into if you like so I can clean up and refactor the code as necessary before doing a master merge.

File: src/ProtoBuf.FSharp/ZeroValues.fs
Comment (mvkara) [inline src/ProtoBuf.FSharp/ZeroValues.fs]: These should not be caps. I will stop making style comments; as per previous comment merge to 'develop' and I will go through it.

File: src/ProtoBuf.FSharp/CodeGen.fs
Comment (mvkara) [inline src/ProtoBuf.FSharp/CodeGen.fs]: Had to read this and the old code a few times to understand the difference in behavior here. In future to save time recommend doing one thing in a PR at a time, it makes it slower to review.

On this part especially around value types:
- I'm guessing your doing this to initialise the value type to a default, since declaring a local does that. I wonder if we can make it more explicit by using Newobj as the opcode? This also creates Value types.

**Review by mvkara — COMMENTED** [review]
Did a quick review. Minus some style comments and questions the only reason I haven't hit "Approve" on the PR is two reasons.

- If you could merge to the "develop" branch I've created I can format the style as appropriate for this repo.
- I can't get the tests to run - the test runner halts. Everytime this has occurred its usually been a bug in my code somewhere. Let me know if you can get a clear test run; otherwise I will take a look. EDIT: Looks like in release configuration it seems to work, in Debug mode it is random whether tests work or fail. May be the tests, I can take a look.

File: src/ProtoBuf.FSharp/CodeGen.fs
Comment (mvkara) [inline src/ProtoBuf.FSharp/CodeGen.fs]: Small question - where is the namespace for the type determined? Normally I would expect it to be here - but you aren't using the "FullName" of the type property. The reason I ask is that it is very common to have duplicate types in schemas (e.g. Types.V1 and Types.V2). We don't want clashes on the generated type name; but I would prefer it to be deterministic as well/repeatable to protect against double registration.

File: src/ProtoBuf.FSharp/CodeGen.fs
Comment (mishun) [reply src/ProtoBuf.FSharp/CodeGen.fs]: That was just a short lazy way which works with all of { structs, primitive types, enums }. If I understand correctly, value types may in principle have parameterless constructor defined, but it's a very exotic thing. Although, flinging uninitialized chunks of memory around as structs may backfire in some spectacular way, so doing OpCodes.Initobj may be a very good idea. And it should be OpCodes.Ldloc instead of OpCodes.Ldloca. I'll fix it.

File: src/ProtoBuf.FSharp/CodeGen.fs
Comment (mishun) [reply src/ProtoBuf.FSharp/CodeGen.fs]: I thought for some reason that name for types emitted with reflection is irrelevant. Apparently that's not true (and in any case it would've appeared in debugger and stack traces)

File: src/ProtoBuf.FSharp/ZeroValues.fs
Comment (mishun) [reply src/ProtoBuf.FSharp/ZeroValues.fs:17]: That develop branch appears to be in different repository

Comment (mvkara) [reply src/ProtoBuf.FSharp/ZeroValues.fs:17]: Fixed. Thanks

**Comment by mishun** [comment]
Tests seem to pass successfully on my machine under both Debug and Release. Maybe it was some undefined behavior related to uninitialized memory?

**Review by mvkara — APPROVED** [review]
Besides a few style points its looking good. Will merge into "Develop" and do some minor fixes and final testing as time permits.

**Comment by mvkara** [comment]
Thanks for the PR.

**Comment by mvkara** [comment]
On this PR I've found some issues caused by this. See PR #15 for introduced record test cases that replicate this. I'm getting segmentation faults (ArchLinux, Net Core 5.0) when attempting to run the tests and/or failing tests. Until I can manage to fix this I can't merge the changes in and release.

**Pull request merged**
